src/base/LruMap.h

```diff
@@ -148,50 +184,70 @@ LruMap<EntryValue, EntryCost>::add(const char *key, EntryValue *t)
     if (memLimit() == 0)
         return false;
 
-    index.push_front(new Entry(key, t));
-    storage.insert(MapPair(key, index.begin()));
+    // optimization: save on memory usage for already-stale objects
+    // (findEntry() would discard instead later anyway)
+    if (ttl < 0)
+        return false;
+
+    auto e = new Entry(key, t, ttl);
+    memUsed_ += e->memoryUsed();
+    index.push_front(e);
+
+    auto m = MapPair(key, index.begin());
+    memUsed_ += memoryUsedBy(m);
+    storage.insert(m);
 
     ++entries_;
-    return true;
+
+    // if over memory limit, start purging
+    trim();
+
+    // the new entry was placed first,
+    // if trim removed it there will be none left.
+    return entries_ >= 1;
 }
 
-template <class EntryValue, size_t EntryCost>
+template <class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
 bool
-LruMap<EntryValue, EntryCost>::expired(const LruMap::Entry &entry) const
+LruMap<EntryValue, MemoryUsedByEV>::expired(const LruMap::Entry &entry) const
 {
-    if (ttl < 0)
+    if (retention < 0)
         return false;
 
-    return (entry.date + ttl < squid_curtime);
+    return (entry.expires < squid_curtime);
```

Comment:
I agree that this is fixed.

> the "date + ttl" part is simply moved from here in expires() to the Entry::Entry() constructor where 'date' is equal to squid_curtime

I may have lost context here, but _if_ you are implying that this problem existed before this PR, then I disagree because the `ttl` in `date + ttl` could not be zero (because it was impossible to cache zero-ttl entries). It is possible to fix this problem by disallowing caching of zero-TTL entries again and your 11575da does that AFAICT.